pkg/resource/deploy/step.go

```diff
@@ -562,6 +579,10 @@ func (s *UpdateStep) Apply(preview bool) (resource.Status, StepCompleteFunc, err
 	return resourceStatus, complete, resourceError
 }
 
+func (s *UpdateStep) Done() {
+	s.reg.Done(&RegisterResult{State: s.new})
```

Comment:
This can't be safe. We haven't actually managed to update to s.new, we should mark this as done for that state.

Reply:
Yeah I'm feeling a bit iffy about this.  It does work, but I'm not sure it's the best way to go, or that it's right.

We did create a `s.new` here, as that's what we initialize the step with, but we didn't fill out any information about the outputs, nor the modified timestamp.  (similarly for `CreateStep` we're not filling in the new ID and not the outputs or timestamps either). Now what done does is unblock the `done` channel [here](https://github.com/pulumi/pulumi/blob/3043adeeb388f31198e0ee42fc708d78c7c0c810/pkg/resource/deploy/source_eval.go#L1864), and we end up returning a valid result to the language SDK.  Meanwhile we are registering the error as mentioned above.

For the current behaviour without `ContinueOnError`, we never close this channel, but we cancel the context of the StepExecutor, so the language SDK just never gets a response back (?? I need to double check that that's what's happing, but I had that impression).  

We could return an error response from `source_eval`, which would also make the `RegisterResource` call above error out, but that makes it throw that error back to the user as well, and we end up with a duplicate error, which I don't think that's the result we want either.  I'll experiment a bit more and will have a look if I can come up with something better.  But any ideas are also welcome.

Reply:
Yeh we have a very similar issue with --targets. 
There's cases where the user program defines a new resource to be created, but we run with targets so it doesn't actually get created but we just return an empty state (I think) back for the resource. I think that causes everything to register as unknown even during a full `up`. 
I don't think any of the SDKs really handle that very well, this looks similar.
I think the right thing to do here is to have a way for register resource to return back that the resource didn't error (at least as far as the user program is concerned) but it also didn't create and so no dependent operations can take place. Might be that's just a case of allowing 'unknown' not just during preview, but during up as well.

Reply:
Ah `--targets` is a good call.  For non-targeted resources there we use a `SameStep`, so the apply actually succeeds, and we re-use the outputs from `old`.  Though I dunno if that makes a lot of sense for updates of failing resources, since the outputs there should really be unknown at that point :thinking:  

Though I'm not even sure we should make the outputs that were previously there for that step unknown, or if we should just return an empty PropertyMap for outputs, since we don't even know if the resource would necessarily have the same outputs as before.

Reply:
I pushed up a potential improvement doing this, handling failures specially, and not returning any outputs the the SDK in that case.

Reply:
In terms of resource application, and what we return to the SDK `SkippedCreateStep` looks like it's the same as `SameStep` to me.  The only difference is that we don't write anything into the snapshot.  

But a failed resource seems kind of different.  I'm not sure we can trust the outputs for those, since we didn't get any back.  For non-targeted resources it seems kind of clear that the outputs haven't changed, but for failed resources this seems like undefined behaviour.  We could parrot back the old outputs, but I'm unsure whether that's better than not returning any outputs, and with that making it clear that the resource is failed and the outputs are not to be trusted?

Reply:
> I'm not sure we can trust the outputs for those, since we didn't get any back.

We didn't get any outputs back for a skipped create either, thus why I think it's an iffy part of the system. Let's not replicate that mistake here.